### PR TITLE
FIX Do not suffix trailing slash to external links

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11151

Since there is behavioral change here, possibly this should target 5 / 5.3, though I'd say it's buggy behaviour so should target 5.2

### CODE REVIEW

File: tests/php/Control/DirectorTest.php [hunk 1]
```diff
@@ -100,30 +100,30 @@ public function testAbsoluteURL()
             //test empty / local urls
             foreach (['', './', '.'] as $url) {
                 $this->assertEquals("http://www.mysite.com:9090/mysite{$slash}", Director::absoluteURL($url, Director::BASE));
-                $this->assertEquals("http://www.mysite.com:9090{$slash}", Director::absoluteURL($url, Director::ROOT));
+                $this->assertEquals("http://www.mysite.com:9090/", Director::absoluteURL($url, Director::ROOT));
```
Comment (GuySartorelli) [inline tests/php/Control/DirectorTest.php]: Why did you change this? And why _didn't_ you change the line above?

File: tests/php/Control/DirectorTest.php [hunk 2]
```diff
@@ -565,6 +565,7 @@ public function providerTestTestRequestCarriesGlobals()
      */
     public function testTestRequestCarriesGlobals($url, $fixture, $method)
     {
+        Controller::config()->set('add_trailing_slash', false);
```
Comment (GuySartorelli) [inline tests/php/Control/DirectorTest.php]: Why did this (and others in this file) get added?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Just a couple questions

File: tests/php/Control/DirectorTest.php (hunk 1 above)
Comment (emteknetnz) [reply tests/php/Control/DirectorTest.php]: It's because of this line

`Director::config()->set('alternate_base_url', 'http://www.mysite.com:9090/mysite/');`

Note that the alternate_base_url is a subdir of mysite.com

This means that the the top-level mysite.com is treated like a completely different website e.g. an external link, so it's excluded from getting the add_trailing_slash logic

File: tests/php/Control/DirectorTest.php (hunk 2 above)
Comment (emteknetnz) [reply tests/php/Control/DirectorTest.php]: Unit tests were failing without it being added

I'll have another look at these ones, there may be a better way to resolve this

Comment (emteknetnz) [reply tests/php/Control/DirectorTest.php]: OK ... It's because I had local config setting it to true :rocket:

File: tests/php/Control/DirectorTest.php (hunk 1 above)
Comment (GuySartorelli) [reply tests/php/Control/DirectorTest.php]: All of these URLs use the alternate base URL, though?

> `$this->assertEquals("http://www.mysite.com:9090/mysite{$slash}"`
> 
> `$this->assertEquals("http://www.mysite.com:9090/"`

These have the exact same base URL as one another, both of which match the `alternate_base_url` domain and shouldn't be seen as external.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: tests/php/Control/DirectorTest.php (hunk 1 above)
Comment (emteknetnz) [reply tests/php/Control/DirectorTest.php]: `$this->assertEquals("http://www.mysite.com:9090/mysite{$slash}"`

The `{$slash}` means it IS being tested that it's adding or removing the slash i.e. it is being treated as the internal (not external) site 

`$this->assertEquals("http://www.mysite.com:9090/"` lacks the `/mysite` suffix - so it's an external site

Comment (GuySartorelli) [reply tests/php/Control/DirectorTest.php]: I've just tested the following:

```php
<?php

use SilverStripe\Control\Director;

Director::config()->set('alternate_base_url', 'http://www.mysite.com:9090/mysite/');
var_dump(Director::is_site_url('http://www.mysite.com:9090/')); // prints bool(true)
```

This shows that `http://www.mysite.com:9090/` is treated as internal by Director. The path is not taken into account.

Comment (GuySartorelli) [reply tests/php/Control/DirectorTest.php]: Stepping through with a debugger, it's being caught by `if ($extension && $url !== Director::config()->get('alternate_base_url')) {` and being incorrectly ignored as a result.

File: src/Control/Controller.php [hunk 3]
```diff
@@ -680,6 +680,18 @@ public static function join_links($arg = null)
...
     {
+        // Do not normalise external urls
+        // Note that urls without a scheme such as www.example.com will be counted as a relative file
+        if (!Director::is_site_url($url)) {
+            return $url;
+        }
+
+        // Do not modify files e.g. some/path/myfile.txt (exclude if $url is absolute_base_url)
+        $extension = pathinfo(rtrim($url), PATHINFO_EXTENSION);
+        if ($extension && $url !== Director::config()->get('alternate_base_url')) {
```
Comment (GuySartorelli) [inline src/Control/Controller.php]:
```suggestion
        if (preg_match('/^(.*)\.([^\/]*)$/', Director::makeRelative($url))) {
```
This seems to work the way we'd expect this to work, and doesn't have the problem in https://github.com/silverstripe/silverstripe-framework/pull/11321#discussion_r1704908697
This regex was what was originally being used to detect files.

This would probably work as well:
```suggestion
        $extension = pathinfo(Director::makeRelative($url), PATHINFO_EXTENSION);
        if ($extension) {
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Control/Controller.php (hunk 3 above)
Comment (emteknetnz) [reply src/Control/Controller.php]: Updated

File: src/Control/Director.php [hunk 4]
```diff
@@ -821,7 +821,11 @@ public static function is_site_url($url)
 
         // Allow extensions to weigh in
         $isSiteUrl = false;
-        static::singleton()->extend('updateIsSiteUrl', $isSiteUrl, $url);
+        // not using static::singleton() here because it can, for unknown reasons, break
+        // functional tests such as those in HTTPCacheControlIntegrationTest
+        // extension hooks will still be called as expected
+        $director = new static();
+        $director->extend('updateIsSiteUrl', $isSiteUrl, $url);
```
Comment (GuySartorelli) [inline src/Control/Director.php]: This screamed "underlying bug" to me so I've looked into it a bit and found the cause.

Instantiating a singleton includes applying the middleware, which I think is ultimately the cause of the failure - but regardless of what it is about the singleton that causes it, the main problem is a singleton of `Director` being instantiated before the test is run.

`Controller::normaliseTrailingSlash()` is called during `SapphireTest::setUp()`
1. `HTTPCacheControlIntegrationTest::setUp()`
1. `FunctionalTest::setUp()`
1. `SapphireTest::setUp()`
1. `SapphireTest::setUpRoutes()`
1. `SapphireTest::getExtraRoutes()`
1. `RequestHandler::Link()`
1. `Controller::join_links()`
1. `Controller::normaliseTrailingSlash()`

We should probably add this to the end of `SapphireTest::setUp()`:

```php
Injector::inst()->unregisterNamedObject(Director::class);
```

This will protect us from similar regressions in the future.

Alternatively, please at least update the comment to reflect that the problem is caused by instantiating a singleton of `Director` prior to tests being run, and that for this specific use case we don't need to use dependency injection because the chance of the `extend()` method being customised in projects is low.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Control/Director.php (hunk 4 above)
Comment (emteknetnz) [reply src/Control/Director.php]: Have gone for the better comment option

**Review by GuySartorelli — APPROVED** [review]
